stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/Messages.kt

```diff
@@ -181,65 +212,160 @@ private fun OnLastVisibleItemChanged(lazyListState: LazyListState, onChanged: (f
-                lazyListState.animateScrollToItem(0)
-            }
+        val shouldScrollToBottomOnNewMessage = shouldScrollToBottomOnNewMessage(
+            focusedItemIndex,
+            firstVisibleItemIndex,
+            newMessageState,
+            areNewestMessagesLoaded,
+            lazyListState.isScrollInProgress
+        )
 
-            !lazyListState.isScrollInProgress && newMessageState == MyOwn -> coroutineScope.launch {
-                if (firstVisibleItemIndex > 5) {
+        if (shouldScrollToBottomOnNewMessage) {
+            coroutineScope.launch {
+                if (newMessageState == MyOwn && firstVisibleItemIndex > 5) {
                     lazyListState.scrollToItem(5)
                 }
                 lazyListState.animateScrollToItem(0)
             }
         }
     }
 
-    if (abs(firstVisibleItemIndex) >= 3) {
+    if (isScrollToBottomButtonVisible(isMessageInThread, firstVisibleItemIndex, areNewestMessagesLoaded)) {
         MessagesScrollingOption(
             unreadCount = messagesState.unreadCount,
             modifier = Modifier.align(Alignment.BottomEnd),
             onClick = {
-                coroutineScope.launch {
-                    if (firstVisibleItemIndex > 5) {
-                        lazyListState.scrollToItem(5)
+                scrollToBottom {
+                    coroutineScope.launch {
+                        if (firstVisibleItemIndex > 5) {
+                            lazyListState.scrollToItem(5)
+                        }
+                        lazyListState.animateScrollToItem(0)
                     }
-                    lazyListState.animateScrollToItem(0)
                 }
             }
         )
     }
 }
 
+/**
+ * Determines if the list should scroll to the bottom when a new message arrives. If we are focusing on an item we do
+ * not wish to take the user off of it, if the newest messages are not loaded then the list will not scroll since the
+ * new message will not be in the list and if the user has scrolled further in search of a certain part of messages
+ * history we should not break that flow.
+ *
+ * @param focusedItemIndex The index of the currently focused item.
+ * @param firstVisibleItemIndex The index of the first visible item in the messages list.
+ * @param newMessageState The [NewMessageState] if a new message has arrived.
+ * @param areNewestMessagesLoaded Whether the newest messages are loaded inside the list or not.
+ * @param isScrollInProgress If the list is currently scrolling or not.
+ *
+ * @return Whether the list should scroll to the bottom when a new message arrives or not.
+ */
+private fun shouldScrollToBottomOnNewMessage(
+    focusedItemIndex: Int,
+    firstVisibleItemIndex: Int,
+    newMessageState: NewMessageState?,
+    areNewestMessagesLoaded: Boolean,
+    isScrollInProgress: Boolean,
+): Boolean {
+    newMessageState ?: return false
+
+    return focusedItemIndex == -1 &&
```

Comment:
Since it is an index and if we search we acquire by `indexOfFirst` if the index is not found `-1` is returned. This check just checks that the focused message is not inside the list.

Reply:
Could we make it into a constant then?

Reply:
We can but then it should be a constant we use for every check if the item is not insde the list. I would leave it as is.